examples/producer_custom_channel_example/producer_custom_channel_example.go

```diff
@@ -0,0 +1,124 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+)
+
+func main() {
+
+	if len(os.Args) != 3 {
+		fmt.Fprintf(os.Stderr, "Usage: %s <bootstrap-servers> <topic>\n",
+			os.Args[0])
+		os.Exit(1)
+	}
+
+	bootstrapServers := os.Args[1]
+	topic := os.Args[2]
+	totalMsgcnt := 3
+
+	p, err := kafka.NewProducer(&kafka.ConfigMap{"bootstrap.servers": bootstrapServers})
+
+	if err != nil {
+		fmt.Printf("Failed to create producer: %s\n", err)
+		os.Exit(1)
+	}
+
+	fmt.Printf("Created Producer %v\n", p)
+
+	// Listen to all the client instance-level errors.
+	// It's important to read these errors too otherwise the events channel will eventually fill up
+	go func() {
+		for e := range p.Events() {
+			switch ev := e.(type) {
+			case kafka.Error:
+				// Generic client instance-level errors, such as
+				// broker connection failures, authentication issues, etc.
+				//
+				// These errors should generally be considered informational
+				// as the underlying client will automatically try to
+				// recover from any errors encountered, the application
+				// does not need to take action on them.
+				fmt.Printf("Error: %v\n", ev)
+			default:
+				fmt.Printf("Ignored event: %s\n", ev)
+			}
+		}
+	}()
+
+	msgcnt := 0
+	for msgcnt < totalMsgcnt {
+		value := fmt.Sprintf("Producer example, message #%d", msgcnt)
+
+		deliveryChan := make(chan kafka.Event)
+		go func() {
+			for e := range deliveryChan {
+				switch ev := e.(type) {
+				case *kafka.Message:
+					m := ev
+					if m.TopicPartition.Error != nil {
+						fmt.Printf("Delivery failed: %v\n", m.TopicPartition.Error)
+					} else {
+						fmt.Printf("Delivered message to topic %s [%d] at offset %v\n",
+							*m.TopicPartition.Topic, m.TopicPartition.Partition, m.TopicPartition.Offset)
+					}
+
+				default:
+					fmt.Printf("Ignored event: %s\n", ev)
+				}
+				// in this case the caller knows that this channel is used only
+				// for one Produce call, so it can close it.
+				close(deliveryChan)
+			}
+		}()
+
+		err = p.Produce(&kafka.Message{
+			TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
+			Value:          []byte(value),
+			Headers:        []kafka.Header{{Key: "myTestHeader", Value: []byte("header values are binary")}},
+		}, deliveryChan)
+
+		if err != nil {
+			close(deliveryChan)
+			if err.(kafka.Error).Code() == kafka.ErrQueueFull {
+				// Producer queue is full, waits for it to be freed
+				time.Sleep(time.Second)
+				continue
+			}
+			fmt.Printf("Failed to produce message: %v\n", err)
+			// Flush and close the producer and the events channel
+			for p.Flush(1000) > 0 {
```

Comment:
I'm not sure we should make this a terminal error, simply printing the error should be sufficient.

Reply:
If people want guaranteed delivery they should the idempotent producer.